Wrappers for Location2D, Direction, EyeTracker2D/3D

JeroMiya opened the pull request:
I don't have any eye tracking hardware to test this with of course, and I still need to figure out how to add direction and location2D plugins to the tree, but I figured it should be ready for a code review at least, which is easiest via a pull request. If you have any suggestions for how to test this (if all else fails, I'll just write some plugins that generate dummy data for each of the new interface types), let me know.

Implements: https://github.com/OSVR/Managed-OSVR/issues/19

JeroMiya committed:
Added interface wrappers for Location2D, Direction, EyeTracker2D, and EyeTracker3D interface types.

rpavlik commented on ClientKit/ClientReportTypes.cs:
```diff
@@ -20,6 +20,15 @@
 using System.Runtime.InteropServices;
 using OSVR.ClientKit;
 
+using ChannelCount = System.Int32;
```
Can we rename this to SensorCount? See https://github.com/OSVR/OSVR-Core/issues/166

JeroMiya replied:
Will do.

rpavlik commented on ClientKit/ClientReportTypes.cs:
```diff
@@ -27,37 +36,81 @@ namespace ClientKit
+        [StructLayout(LayoutKind.Sequential)]
+        public struct EyeTracker3DState
+        {
+            public bool directionValid;
```
I know these need to be like this for pinvoke, but if there's a nice .NET-native way of doing something like `boost::optional` does for C++ - a member that may or may not have a value - that would be a good thing to use.

JeroMiya replied:
There is Nullable<T> or just boxed value types, but not for marshalling structs like these, as far as I know. Plus they would cause GC pressure even if you could.

rpavlik replied:
Is there a solution that could be used post-marshalling, like the callback to event stuff, that wouldn't cause gc pressure? (It's OK to say no :) ) Just curious.

JeroMiya replied:
Well, we could override the `OnStateChanged` method in `InterfaceBase` to filter out reports where there is no value, but that only works when there is only one "optional" field in the report type, and it wouldn't change what is returned by `GetState`. The only way to avoid GC pressure would be some kind of managed ring-buffer or memory pooling for the boxed values, which would need to be synchronized (not that the managed interfaces are thread safe yet anyway).

rpavlik replied:
OK, ugh, not worth it. Users can just look at the bools like everybody else then. Was just curious since we'll be adding a new tracker callback type that has bools as well, to cleanly handle the case where an app wants full pose if it can get it, otherwise it's ok with some subset - the so-called "subset message" problem.

JeroMiya committed:
Renamed ChannelCount to SensorCount.

JeroMiya committed:
Added code to the Direction, EyeTracker*, and Location2D samples to test the state APIs.

JeroMiya committed:
Changed the osvr path for the Location2D and Direction samples to match what is exposed by default for the com_osvr_example_EyeTracker plugin. Also added a comment with instructions for how to get the sample eye-tracker plugin working for the example.

JeroMiya commented:
I was able to test all of the new interfaces using the sample com_osvr_example_EyeTracker from the OSVR-Core source. I added some code to the samples to exercise the state APIs, and added instructions in comments for getting that plugin to work. I also adjusted the sample osvr paths for that plugin. This should be ready to merge now.

rpavlik commented:
Cool, thank you. I hate to keep tacking on things, but skimming the diff I noticed that we're marshalling blink state as a byte - which is correct, by the API, but semantically it's actually a bool (0 false, 1 true). Is that easy to add to the pinvoke code? (to interact with the native code using a byte, but expose to managed code as a bool) Just trying to avoid people writing code against stuff that we'll change shortly, or I'd put this separately. If it is possible, would be nice to do for blink state here, and then we can open an issue for adjusting button to be the same, though that (I think?) will potentially introduce a break.

JeroMiya committed:
EyeTrackerBlinkReport.state is now marshalled as a bool instead of a byte in both the callback report type and the state API.

JeroMiya commented:
OK, I changed the blink interface to marshall the byte value as a boolean .The `[MarshallAs(UnmanagedType.I1)]` attribute on both the report field and the state API argument is important, otherwise the bool is expanded to 4 bytes. A value of 0 is marshalled as 'false' and anything else is marshalled as 'true' (so signed/unsigned shouldn't matter).

Reference: http://blogs.msdn.com/b/jaredpar/archive/2008/10/14/pinvoke-and-bool-or-should-i-say-bool.aspx

I could do the same for the button interface, but as you say this would be a breaking change.

Question: does the blink interface actually track a blink as a state, as in "your eyes are closed, now they're open", or is it detecting a distinct "blink" event, emulated as a state (one report of true for the detected 'blink', and all other reports false)? Or is that an implementation detail of the interface plugin/hardware?

rpavlik commented:
In theory it's actually tracked as a state, and I think most hardware recognizes it as such (since it can't really track your pupils when your eyes are closed) so while an occasional plugin might do a close then open immediately, in general that's not the case.
